### PR TITLE
[WIP] Implement transaction email services

### PR DESCRIPTION
**Pull request by kimlisa**

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -107,21 +112,54 @@ func (s *Service) CreateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
...
+		logger.Error("CreateUser INSERT user_svc.accounts:", err.Error())
+		return nil, status.Error(codes.Unknown, err.Error())
+	}
+	logger.Info("Success INSERT new user:", user.GetFirstName(), user.GetLastName(), user.GetUuid())
+
+	// insert token into db
+	command = `INSERT INTO user_svc.pending_tokens(token, created_date, uuid) VALUES($1, $2, $3)`
+	_, err = postgresDB.Exec(command, token, time.Now().UTC(), user.GetUuid())
+	if err != nil {
+		logger.Error("CreateUser INSERT user_svc.pending_tokens:", err.Error())
```
Comment (kimlisa) [inline service/service.go]: starting from this block on and onwards (refer to the images I posted), its a huge chunk I can't test, so I am wondering if I should move this to a utility or db function so I can test that instead.

Comment (faraonc) [reply service/service.go]: I figure that we can add a task to make utility functions that generate query strings.

File: service/email.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	if err := r.parseTemplates(filePaths); err != nil {
+		return err
+	}
+
+	if err := r.processEmail(); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (faraonc) [inline service/email.go]: Since sending an email will be a common operation, we need a high-level documentation on how you are doing this process.

File: service/email.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+const (
+	// MIME (Multipurpose Internet Mail Extension), extends the format of email
+	mime                = "MIME-version: 1.0;\nContent-Type: text/html; charset=\"UTF-8\";\n\n"
+	subjectVerifyEmail  = "Verify account for Humpback whale Social Call"
+	templateVerifyEmail = "verify_email.html"
+)
+
+func newEmailRequest(data map[string]string, to []string, from string, subject string) (*emailRequest, error) {
```
[the remark this answers is not in the document]
Comment (faraonc) [reply service/email.go]: Should `data` not be nil?

File: service/email.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	// note, data can be nil
+	if to == nil || from == "" || subject == "" {
+		return nil, errEmailRequestFieldsEmpty
+	}
+
```
Comment (faraonc) [inline service/email.go]: Emails can be checked, i know that is possible in Java, but not sure about in Golang. There is a package called https://github.com/badoux/checkmail

File: service/email.go (hunk 2 above)
Comment (faraonc) [reply service/email.go]: I am assuming you used this reference https://www.ardanlabs.com/blog/2013/06/send-email-in-go-with-smtpsendmail.html

File: service/email.go (hunk 4 above)
Comment (faraonc) [reply service/email.go]: I am not sure if we should let gmail handle the checking for us

File: service/email_test.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+			assert.NotNil(t, err)
+		} else {
+			assert.Nil(t, err)
+			assert.Equal(t, c.expMsg, "")
+		}
+	}
+}
+
+func TestSendEmail(t *testing.T) {
+	email := []string{"hwsc.test+user0@gmail.com"}
```
Comment (faraonc) [inline service/email_test.go]: we need a mock to test email being sent http://tmichel.github.io/2014/10/12/golang-send-test-email/

File: service/service.go (hunk 1 above)
Comment (faraonc) [reply service/service.go]: Add a `// TO-DO https://github.com/hwsc-org/hwsc-user-svc/issues/21`

File: service/service.go [hunk 6]
```diff
@@ -107,21 +112,54 @@ func (s *Service) CreateUser(ctx context.Context, req *pb.UserRequest) (*pb.User
 	user.Password = hashedPassword
 	user.IsVerified = false
 
-	// insert into DB
+	// create unique email token
+	token, err := generateEmailToken()
```
Comment (faraonc) [inline service/service.go]: What if users use similar e-mails?

File: tmpl/header.tmpl [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            body {
+                margin: 0 auto;
+                padding: 0;
+            }
+            .small-print {
+                font-size: 12px;
+                color: #656565;
+            }
+            .button-container {
+                padding: 20px 0 40px 0;
```
Comment (faraonc) [inline tmpl/header.tmpl]: can we use relative units?

**Comment by faraonc** [comment]
Below are two additional issues added.

RPC service to resend email verification
https://github.com/hwsc-org/hwsc-user-svc/issues/22
db.go query generator
https://github.com/hwsc-org/hwsc-user-svc/issues/21

File: service/email.go (hunk 4 above)
Comment (kimlisa) [reply service/email.go]: I can actually use this package to check email validity before inserting the new user into the database.

Do you think I will still need to check if email exist before sending any kind of email?

File: service/email_test.go (hunk 5 above)
Comment (kimlisa) [reply service/email_test.go]: I read this article couple times... but I am confused what the purpose of this is? Am I not doing it already?

File: service/service.go (hunk 6 above)
Comment (kimlisa) [reply service/service.go]: I will move this block after I insert the new user.
Duplicate email will be catched by the db as unique violation error, and will return the duplicate email error.

File: service/email.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kimlisa) [reply service/email.go]: data can be nil if the templates contain only static data

File: service/email.go (hunk 4 above)
Comment (kimlisa) [reply service/email.go]: And we don't have to use gmail, in fact using gmail with plainAuth is an issue because it forces me to enable "less secure apps" which isn't secure atm. But for testing purposes it serves well.

File: service/email.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	to           []string
+	subject      string
+	body         string
+	templateData map[string]string
+}
+
+const (
+	// MIME (Multipurpose Internet Mail Extension), extends the format of email
+	mime                = "MIME-version: 1.0;\nContent-Type: text/html; charset=\"UTF-8\";\n\n"
+	subjectVerifyEmail  = "Verify account for Humpback whale Social Call"
```
[the remark this answers is not in the document]
Comment (faraonc) [reply service/email.go]: Is this intentionally lower case?

File: service/email.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	if filePaths == nil {
+		return errEmailNilFilePaths
+	}
+
+	templates, err := template.ParseFiles(filePaths...)
+	if err != nil {
+		return err
+	}
+
+	buffer := *new(bytes.Buffer)
```
Comment (faraonc) [inline service/email.go]: `buffer := &bytes.Buffer{}` ?

File: service/email_test.go (hunk 5 above)
Comment (faraonc) [reply service/email_test.go]: we can write an issue for this, unless we are entirely sure what email service are we using.

File: service/service_test.go [hunk 10]
```diff
@@ -44,45 +44,70 @@ func TestGetStatus(t *testing.T) {
...
+	}
+
+	// fail: passwords with leading/trailing spaces
+	testUser6 := &pb.User{
+		FirstName: "Lisa",
+		LastName:  "Kim",
+		Email:     "hwsc.test+user3@gmail.com",
+		Password:  "    abcklajdsfasdf      ",
+	}
+
```
Comment (faraonc) [inline service/service_test.go]: Missing tests for zero values like empty strings or nil for other properties.

File: service/utility.go [hunk 11]
```diff
@@ -34,7 +36,7 @@ func validateUser(user *pb.User) error {
 	if err := validateEmail(user.GetEmail()); err != nil {
 		return err
 	}
-	if user.GetPassword() == "" {
+	if pass := user.GetPassword(); pass == "" || strings.TrimSpace(pass) != pass {
```
Comment (faraonc) [inline service/utility.go]: I do not agree with the var name `pass` for password, but if you have a reason then I would like to know.

File: service/utility.go [hunk 12]
```diff
@@ -101,16 +103,31 @@ func generateUUID() (string, error) {
 	return strings.ToLower(id.String()), nil
 }
 
-// hashPassword hashes and salts provided string
+// TODO probably need to synchronize this as well
```
Comment (faraonc) [inline service/utility.go]: need an issue to track this so we do not forget

File: tmpl/header.tmpl (hunk 7 above)
Comment (faraonc) [reply tmpl/header.tmpl]: any discussion?

Comment (kimlisa) [reply tmpl/header.tmpl]: I did use relative units...

File: service/email.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kimlisa) [reply service/email.go]: What? I changed it...

File: tmpl/header.tmpl (hunk 7 above)
Comment (faraonc) [reply tmpl/header.tmpl]: https://www.tutorialrepublic.com/css-tutorial/css-units.php

File: service/utility.go (hunk 12 above)
Comment (kimlisa) [reply service/utility.go]: https://github.com/hwsc-org/hwsc-user-svc/issues/24
